rfcs/0049-flakes.md

```diff
@@ -662,6 +660,17 @@ simpler format such as JSON or TOML. This avoids the Turing tarpit
 where getting flake metadata requires the execution of an arbitrarily
 complex, possibly non-terminating program.
 
+Flakes could be implemented as an external tool on top of Nix. Indeed,
+there is nothing that flakes allow you to do that couldn't previously
+be done using `fetchGit`, the `--pure-eval` flag and some shell
+scripting. However, implementing flake-like functionality in an
+external tool would defeat the goals of this RFC. First, it probably
```

Comment:
What's the reasoning behind Flakes deserving to be blessed as *the* standard before they've proven their merit or even having given other implementations a chance? Or is the intention to race RFCs in becoming the standard?

I like the Flakes RFC btw. I just don't see why it should be treated differently.

Reply:
> What's the reasoning behind Flakes deserving to be blessed as the standard before they've proven their merit or even having given other implementations a chance?

Well, no Nix feature has ever followed that process before being included, except `nix-repl` (and that was only because I didn't want an ncurses dependency in Nix at the time).

I'm very skeptical that external implementations would ever end up being merged in practice. See `yum`, which is still separate from `rpm` after 20 years.

Reply:
Counterexamples: Cargo, `go mod`.

Reply:
If anything, I think for a shift such as the one this RFC suggests, trying it out in practice somewhat at scale would make it *easier* to adopt it.  That way, potential issues in the design can be revealed and iterated on/addressed *before* we bolt the design down.

I'm not really familiar with the rpm ecosystem, so I don't know why `rpm` and `yum` are separate tools. It might in part just boil down to historical reasons though. 

I'm not sure I see why the (to me fairly marginal) risk of having two tools instead of one is bigger than the risk of the community moving in a direction and then looking back in retrospect and wanting to make changes to the flake design.

Don't get me wrong, I like a lot in the Flakes design. I just don't want us to have to repeat this process in a few years' time just because we didn't test the design well enough before giving it its blessing.

Reply:
Given the current revision of this RFC to only talk about the flake format, this paragraph seems out of place.

Reply:
I’m not sure it is. Whether it would be possible to generate this lockfile using an out-of-tree tool, or has to be implemented in Nix itself, is an important thing to know.

Reply:
I don't know the full history, but I believe families like apt/deb, yum/rpm/dnf, gem/bundler were often built separately so as not to risk disruption by drastically modifying the (legacy) original tool, or to add on extra layers of functionality which only a subset of the community wanted (at the time).

If external-flakes is built with the explicit goal that it will be modified until it's either made part of nix or abandoned, that seems like it would alleviate that risk perfectly well.

I for one would love to try out a real implementation before comitting to it, since it's much harder to spot some usability or functionality gaps in an RFC than in practice. And I have a bunch of my own projects which are ripe for trying out a distributed approach to nix expressions.

Reply:
> I'm very skeptical that external implementations would ever end up being merged in practice. See yum, which is still separate from rpm after 20 years.

Technically speaking, Nix and Nixpkgs are still separate repositories, and there are some cases where it causes a need for more complicated coordination. Still, Nixpkgs is the defacto repository of packages for Nix, but there are forks that are usable by anyone who is interested.

At the same time, NixOS _was_ a separate repository at some point, then got into the same monorepo. So merges happen here if they make sense.

What is the specific problem being addressed by flakes being built-in and not just the defacto tool?